File: Leetcode Challenge/07_July_2020/Python/Week 2/4_Subsets.py
#Given a set of distinct integers, nums, return all possible subsets (the power set).
#
#Note: The solution set must not contain duplicate subsets.
#
#Example:
#
#Input: nums = [1,2,3]
#Output:
#[
#  [3],
#  [1],
#  [2],
#  [1,2,3],
#  [1,3],
#  [2,3],
#  [1,2],
#  []
#]

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        '''
        self.ans = []
        self.dfs([-1], nums)
        return self.ans
    
    def dfs(self, current, nums):
        self.ans.append([nums[i] for i in current][1:])
        for i in range(current[-1] + 1, len(nums)):
            self.dfs(current + [i], nums)
        '''
        '''
        ans = [[]]
        for num in nums:
            ans += [n + [num] for n in ans]
        return ans
        '''
        n = len(nums)
        ans = []
        for i in range(1 << n, 1 << (n + 1)):
            bm = bin(i)[3:] # bitmask
            ans.append([nums[j] for j in range(n) if bm[j] == '1'])
        return ans
        

File: Leetcode Challenge/07_July_2020/Rust/Week 2/4_Subsets.rs
//Given a set of distinct integers, nums, return all possible subsets (the power set).
//
//Note: The solution set must not contain duplicate subsets.
//
//Example:
//
//Input: nums = [1,2,3]
//Output:
//[
//  [3],
//  [1],
//  [2],
//  [1,2,3],
//  [1,3],
//  [2,3],
//  [1,2],
//  []
//]

impl Solution {
    pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
        
        let mut ans : Vec<Vec<i32>> = Vec::new();
        ans.push(Vec::new());
        let mut n;
        for num in nums {
            let mut temp : Vec<Vec<i32>> = Vec::new();
            for N in &ans {
                n = N.clone();
                n.push(num);
                temp.push(n);
            }
            ans.append(&mut temp);
        }
        return ans;
        
        /*
        let n = nums.len();
        let mut ans : Vec<Vec<i32>> = Vec::new();
        for i in (1 << n)..(1 << (n + 1)) {
            let mut bm : Vec<i8> = Vec::new();
            let mut no : i32 = i as i32;
            while no > 0 {
                bm.push((no & 1) as i8);
                no >>= 1;
            }
            let mut t : Vec<i32> = Vec::new();
            for j in 0..n {
                if bm[j] == 1 {
                    t.push(nums[j]);
                }
            }
            ans.push(t);
        }
        return ans;
        */
    }
}